kernel/src/snapshot.rs

```diff
@@ -321,29 +321,64 @@ fn read_last_checkpoint(
 
 /// List all log files after a given checkpoint.
 fn list_log_files_with_checkpoint(
-    cp: &CheckpointMetadata,
+    checkpoint_metadata: &CheckpointMetadata,
     fs_client: &dyn FileSystemClient,
     log_root: &Url,
 ) -> DeltaResult<(Vec<ParsedLogPath>, Vec<ParsedLogPath>)> {
-    let version_prefix = format!("{:020}", cp.version);
+    let version_prefix = format!("{:020}", checkpoint_metadata.version);
     let start_from = log_root.join(&version_prefix)?;
 
-    let files: Vec<ParsedLogPath> = fs_client
-        .list_from(&start_from)?
-        .flat_map(|file| file.and_then(ParsedLogPath::try_from).transpose())
-        .try_collect()?;
-    let (mut commit_files, checkpoint_files): (Vec<_>, _) = files
-        .into_iter()
+    let mut max_checkpoint_version = checkpoint_metadata.version;
+    let mut checkpoint_files = vec![];
+    let mut commit_files = Vec::with_capacity(10);
+
+    for meta_res in fs_client.list_from(&start_from)? {
+        let meta = meta_res?;
+        let parsed_path = ParsedLogPath::try_from(meta)?;
         // TODO this filters out .crc files etc which start with "." - how do we want to use these kind of files?
-        .filter(|p| p.is_commit() || p.is_checkpoint())
-        .partition(ParsedLogPath::is_commit);
+        if let Some(parsed_path) = parsed_path {
+            if parsed_path.is_commit() {
+                commit_files.push(parsed_path);
+            } else if parsed_path.is_checkpoint() {
+                match parsed_path.version.cmp(&max_checkpoint_version) {
+                    Ordering::Greater => {
+                        max_checkpoint_version = parsed_path.version;
+                        checkpoint_files.clear();
+                        checkpoint_files.push(parsed_path);
+                    }
+                    Ordering::Equal => checkpoint_files.push(parsed_path),
+                    Ordering::Less => {}
+                }
+            }
+        }
+    }
+
+    if checkpoint_files.is_empty() {
+        // TODO: We could potentially recover here
```

Comment:
We could certainly create the issue, but IMO it's extra complexity to support a broken use case -- the situation can only arise if somebody deleted files from the delta log directory. Once we start trying to handle those scenarios it's a bottomless pit because literally anything can happen in any order.